text/0000-try-trait-v2.md

```diff
@@ -0,0 +1,970 @@
+```
+
+If you've used `?`-on-`Result` before, that output type is likely unsurprising.  Since it's given out directly from the operator, there's not much of a choice.
+
+The residual types, however, are somewhat more interesting.  Code using `?` doesn't see them directly -- their usage is hidden inside the desugaring -- so there are more possibilities available.  So why are we using these ones specifically?
+
+Most importantly, this gives each family of types (`Result`s, `Option`s, `ControlFlow`s) their own *distinct* residual type.  That avoids unrestricted *interconversion* between the different types, the ability to arbitrarily mix them in the same method.  For example, it was mentioned earlier that just because a `ControlFlow::Break` is also an early exit, that doesn't mean that it should be allowed to consider it a `Result::Err` -- it might be a success, conceptually.  So by giving `ControlFlow<X, _>` and `Result<_, X>` different residual types, it becomes a compilation error to use the `?` operator on a `ControlFlow` in a method which returns a `Result`.  (There are also ways to allow interconversion is specific situations where it's desirable.)
+
+> 🏗️ Note for those familiar with the previous RFC 🏗️
+>
+> This is the most critical semantic difference.  Structurally this definition of the trait is very similar to the previous -- there's still a method splitting the type into a discriminated union between two associated types, and constructors to rebuild it from them.  But by keeping the "result-ness" or "option-ness" in the residual type, it gives extra control over interconversion that wasn't possible before.  The changes other than this are comparatively minor, typically either rearrangements to work with that or renamings to change the vocabulary used in the trait.
+
+Using `!` is then just a convenient yet efficient way to create those residual types.  It's nice as a user, too, not to need to understand an additional type.  Just the same "it can't be that one" pattern that's also used in `TryFrom`, where for example `i32::try_from(10_u8)` gives a `Result<i32, !>`, since it's a widening conversion which cannot fail.  Note that there's nothing special going on with `!` here -- any uninhabited `enum` would work fine.
+
+
+## How error conversion works
+
+One thing [The Book mentions](https://doc.rust-lang.org/stable/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator),
+if you recall, is that error values in `?` have `From::from` called on them, to convert from one error type to another.
+
+The previous section actually lied to you slightly: there are *two* traits involved, not just one.  The `from_residual` method is on `FromTryResidual`, which is generic so that the implementation on `Result` can add that extra conversion.  Specifically, the trait looks like this:
+
+```rust
+trait FromTryResidual<Residual = <Self as Try>::Residual> {
```

Comment:
Yeah, it'll error on use -- if you try to `impl FromTryResidual for i32` it'll give the "it's not `Try`" error, but `impl FromTryResidual<Bar> for i32` won't error because then the default is unused.

Reply:
I didn't *see* any example in the RFC of an `impl FromTryResidual` for a type that doesn't also implement `Try`. It sounds from the dialogue here that you do intend for people to do that, and such usage is what I imagine motivates making it a super-trait (rather than just having a `fn from_residual` method in the `Try` trait itself).

But I think the text of the RFC would benefit from an explicit example of where you expect `FromTryResidual` to be implemented on a type that does not implement `Try`.

Reply:
Or ... I guess the `impl` below of `impl<T, E: From<FancyError>> FromTryResidual<ResultCodeResidual> for Result<T, E>` is meant to serve as that example? Its not that `Result` doesn't implement `Try`, but rather you now enable this kind of type-specific residual handling without having to reimplement the rest of the try interface?

I guess I need that spelled out a bit more explicitly, in that I would like to know which of the two cases I outlined is the *primary* intent here.